src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs

```diff
@@ -687,26 +689,64 @@ internal void TogglePreviewControlAllowance()
+        /// Because nodes' ContextMenus are loaded lazily, and their MenuItems are disposed on closing,
+        /// these custom MenuItems need to be manually re-injected into the context menu whenever it is opened.
+        /// </summary>
+        private void StashNodeViewCustomizationMenuItems()
         {
-            Guid nodeGuid = ViewModel.NodeModel.GUID;
-            ViewModel.DynamoViewModel.ExecuteCommand(
-                new DynCmd.SelectModelCommand(nodeGuid, Keyboard.Modifiers.AsDynamoType()));
+            foreach (var obj in grid.ContextMenu.Items)
+            {
+                if (!(obj is MenuItem menuItem)) continue;
 
-            viewModel.OnSelected(this, EventArgs.Empty);
-            grid.ContextMenu.DataContext = viewModel;
-            grid.ContextMenu.IsOpen = true;
+                // We don't stash default MenuItems, such as 'Freeze'.
+                if (NodeContextMenuBuilder.NodeContextMenuDefaultItemNames.Contains(menuItem.Header.ToString())) continue;
+
+                // We don't stash the same MenuItem multiple times.
+                if (NodeViewCustomizationMenuItems.ContainsKey(menuItem.Header.ToString())) continue;
+                
+                // The MenuItem gets stashed.
+                NodeViewCustomizationMenuItems.Add(menuItem.Header.ToString(), menuItem);
+            }
         }
 
-        private void topControl_MouseRightButtonDown(object sender, MouseButtonEventArgs e)
+        /// <summary>
+        /// A common method to handle the node Options Button being clicked and
+        /// the user right-clicking on the node body to open its ContextMenu.
+        /// </summary>
+        private void DisplayNodeContextMenu(object sender, RoutedEventArgs e)
         {
             Guid nodeGuid = ViewModel.NodeModel.GUID;
             ViewModel.DynamoViewModel.ExecuteCommand(
                 new DynCmd.SelectModelCommand(nodeGuid, Keyboard.Modifiers.AsDynamoType()));
 
-            grid.ContextMenu.DataContext = viewModel;
-            grid.ContextMenu.IsOpen = true;
+            ContextMenu contextMenu = grid.ContextMenu;
+
+            // Stashing any injected MenuItems from the Node View Customization process.
+            if (contextMenu.Items.Count > 0 && NodeViewCustomizationMenuItems.Count < 1)
+            {
+                StashNodeViewCustomizationMenuItems();
+            }
+
+            // Clearing any existing items in the node's ContextMenu.
+            contextMenu.Items.Clear();
+            NodeContextMenuBuilder.Build(contextMenu, viewModel, NodeViewCustomizationMenuItems);
             
+            contextMenu.DataContext = viewModel;
+            contextMenu.IsOpen = true;
+
+            e.Handled = true;
+        }
+
+        private void MainContextMenu_OnClosed(object sender, RoutedEventArgs e)
+        {
+            grid.ContextMenu.Items.Clear();
```

Comment:
if you've already gone this far, can't we get rid of the ContextMenu as well, why instantiate one contextMenu per nodeView when we only need 1 at a time.

Reply:
If it's possible to style this centrally then it should be possible. However, we're relying on a lot of locally-defined XAML styles currently.